### PR TITLE
rustc: move debug info from LocalDecl and UpvarDecl into a dedicated VarDebugInfo.

### PR DESCRIPTION
**Pull request by eddyb**
This PR introduces a MIR "user variable" debuginfo system, which amounts to mapping a variable name, in some `SourceScope`, to a `Place`, so that:

* each name can appear multiple times (e.g. due to macro hygiene), even in the same scope
* each `Place` can appear multiple times (e.g. in the future from optimizations like NRVO, which collapse multiple MIR locals into one)
* the `Place`s aren't limited to just locals, so they can describe the (right now quite ad-hoc) closure upvars and generator saved state fields, and can be properly transformed by optimizations (e.g. inlining - see `src/test/mir-opt/inline-closure-captures.rs`)

The main motivation for this was that #48300 and further optimizations were blocked on being able to describe complex debuginfo transformations (see https://github.com/rust-lang/rust/pull/48300#discussion_r170020762).

<hr/>

In the textual representation, the "user variable" debuginfo can be found in each scope, and consists of `debug NAME => PLACE;` "declarations", e.g. the MIR for `let x = ...; let y = ...; ...` is now:
```rust
    let _1: T;                           // in scope 0 at ...
    scope 1 {
        debug x => _1;                   // in scope 1 at ...
        let _2: T;                       // in scope 1 at ...
        scope 2 {
            debug y => _2;               // in scope 2 at ...
        }
    }
```
For reference, this is how the information was represented before this PR:
(notably, the scopes in which the variables are visible for debuginfo weren't even shown anywhere, making `scope 2` look pointless, and user variable names were part of MIR locals)
```rust
    let _1: T;                           // "x" in scope 0 at ...
    scope 1 {
        let _2: T;                       // "y" in scope 1 at ...
        scope 2 {
        }
    }
```

cc @nikomatsakis @michaelwoerister

### CODE REVIEW

**Comment by Dylan-DPC** [comment]
ping from triage @eddyb @matthewjasper what's the update on this? is this blocked on anything?

**Comment by eddyb** [comment]
Blocked on #56278. Also, r? @nikomatsakis

**Comment by cuviper** [comment]
> when, if ever, can we bump minimum LLVM to 6?

See #56642.

**Comment by nikomatsakis** [comment]
Closing in favor of https://github.com/rust-lang/rust/pull/56278 for now

**Comment by eddyb** [comment]
I've reopened this, but note that tests won't pass for a while (as the last/only commit isn't finished).

File: src/test/mir-opt/inline-closure-borrows-arg.rs
Comment (eddyb) [inline src/test/mir-opt/inline-closure-borrows-arg.rs]: This is wrong, it should point to the same variable `debug r` does (thanks, @github, for eating my more detailed comment).
I think this might just be copyprop getting rid of the `variable = r` assignment without knowing about debuginfo.

File: src/librustc_codegen_ssa/mir/analyze.rs
Comment (eddyb) [reply src/librustc_codegen_ssa/mir/analyze.rs]: I guess if you had further projections? Not sure what I meant tbh.

File: src/librustc_codegen_ssa/mir/analyze.rs
Comment (eddyb) [reply src/librustc_codegen_ssa/mir/analyze.rs]: I was staring at that earlier and I'm not sure what should be written there  tbh.

File: src/librustc_mir/borrow_check/mod.rs
Comment (eddyb) [reply src/librustc_mir/borrow_check/mod.rs]: Yeah, and borrowck runs before optimizations.
OTOH, this *is* a complete hack, cc @matthewjasper.

File: src/librustc_mir/borrow_check/mod.rs
Comment (eddyb) [reply src/librustc_mir/borrow_check/mod.rs]: Can we even perform that optimization without dedicated analysis? (It's what you get from my NRVO optimization, I think, perhaps with the SROA one too).

And we would care, because MIR borrowck needs to report errors about user variables (which `a` and `b` are).

The only non-local debuginfo during MIR borrowck, right now, are upvars.

File: src/librustc/mir/mod.rs
Comment (eddyb) [reply src/librustc/mir/mod.rs]: Oh how I love getting reminded `/j/` exists and is distinct from `/i/`:rage:

File: src/librustc_codegen_ssa/mir/analyze.rs
Comment (tmandry) [inline src/librustc_codegen_ssa/mir/analyze.rs]: This comment needs updating or should be removed.

**Comment by Centril** [comment]
@eddyb What is the status of this? Should we close until you can get back to it?

File: src/librustc_codegen_llvm/debuginfo/mod.rs
Comment (oli-obk) [inline src/librustc_codegen_llvm/debuginfo/mod.rs]: The original code was passing a zero length array for direct variables, what's the reason for this change?

File: src/librustc_codegen_ssa/mir/debuginfo.rs
Comment (Centril) [reply src/librustc_codegen_ssa/mir/debuginfo.rs]: See also https://github.com/rust-lang/rust/pull/49229 -- you might be able to nix `DisplayViaDebug`.

File: src/librustc_codegen_llvm/debuginfo/mod.rs
Comment (eddyb) [reply src/librustc_codegen_llvm/debuginfo/mod.rs]: It was passing a zero-length array of "operations", but now it would make more sense to always have the "immediate/direct" offset, and requiring that a slice argument is never empty feels wrong.

It's also easier to explain that "there is one indirection per `indirect_offsets` element" than "there are `offsets.len() - 1` indirections".

File: src/librustc_codegen_ssa/mir/debuginfo.rs
Comment (eddyb) [reply src/librustc_codegen_ssa/mir/debuginfo.rs]: I actually plan to get rid of `DisplayViaDebug` by simply always creating the `String` and just wrapping name-setting logic in `if !cx.sess().fewer_names()`.

**Comment by eddyb** [comment]
I've opened #65718 for the codegen side of this, and rebased this PR on it.
Review comments for all but the last commit of this PR should be made on #65718.

**Comment by eddyb** [comment]
@bjorn3 Please leave comments on the commit in the PR list, not on the commit itself, because I can't reply to the latter from the PR view, and they might even disappear.

**Comment by bjorn3** [comment]
I didnt know there was a difference between selecting a commit from the PR commit list and selecting it from the push message in the PR discussion. Will use the PR commit list next time.

**Comment by eddyb** [comment]
Push message appears to be the problem here, the lines just before the push message have e.g.:
```
https://github.com/rust-lang/rust/pull/56231/commits/5ae8ba83c2f0964372a4777347711ca82785b538
```
which will select the commit but provide the PR interface.

**Comment by bjorn3** [comment]
This is a mobile vs desktop issue. On mobile clicking on a push message goes to the raw commit, on desktop it goes to the commit with PR interface.

**Comment by eddyb** [comment]
This is increasingly offtopic, but for me even on desktop this message :
> @eddyb force-pushed the eddyb:mir-debuginfo branch from c98c310 to 5ae8ba8 1 hour ago

links to those commits directly, not to the PR view of them.

**Comment by bjorn3** [comment]
Same for me. I was actually refering to the commit messages shown just above `@eddyb force-pushed ...`.

**Comment by oli-obk** [comment]
Needs a rebase and there are a bunch of unresolved review comments still left

**Comment by JohnCSimon** [comment]
Ping from triage: 
@eddyb Can you please address the merge conflicts and the unresolved review comments? Thank you!
cc: @oli-obk @cuviper

**Comment by JohnCSimon** [comment]
Pinging  from triage:
@eddyb Can you please address the merge conflicts and the unresolved review comments? Thank you!

File: src/librustc_codegen_ssa/mir/analyze.rs
Comment (eddyb) [reply src/librustc_codegen_ssa/mir/analyze.rs]: There is still a recursive call to `process_place` (sadly), so this appears to still be correct.

File: src/librustc_codegen_ssa/mir/analyze.rs
Comment (eddyb) [reply src/librustc_codegen_ssa/mir/analyze.rs]: I think the "it" is from a truncated "it's possible" (or thereabouts).

File: src/librustc_codegen_ssa/mir/analyze.rs
Comment (eddyb) [reply src/librustc_codegen_ssa/mir/analyze.rs]: Oh I think I literally meant "all cases but debuginfo".

File: src/librustc_codegen_llvm/debuginfo/metadata.rs
Comment (oli-obk) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]: Keep it in a separate function taking just a tcx and a defid

File: src/librustc_codegen_ssa/mir/analyze.rs
Comment (oli-obk) [inline src/librustc_codegen_ssa/mir/analyze.rs]: We are at the slice of projections representation. So not have to but should

**Review by oli-obk — CHANGES_REQUESTED** [review]
r=me with generator stuff pulled in a function

**Comment by eddyb** [comment]
Oh, one more thing I shouldn't forget about, especially once this PR lands:
I need to update https://rust-lang.github.io/rustc-guide/mir/index.html to include information about the new "user variable" debuginfo in MIR (similar to PR description, but expanded a bit).

I should also search for ` // "\w+" in scope ` because that's likely a MIR snippet with an old-style variable name, so it should be updated (this also will showcase the new `debug` declarations).

**Pull request merged**
